contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,316 @@
+
+  Product[] public products;
+  mapping(uint => uint) capacityFactors;
+
+  mapping(uint => uint) initialPrices;
+  mapping(uint => mapping(address => uint)) lastPrices;
+  mapping(uint => mapping(address => uint)) lastPriceUpdate;
+
+
+  /* === CONSTANTS ==== */
+
+  address constant ETH = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+  uint public REWARD_BPS = 5000;
+  uint public constant PERCENTAGE_CHANGE_PER_DAY_BPS = 100;
+  uint public constant BASIS_PRECISION = 10000;
+  uint public constant STAKE_SPEED_UNIT = 100000e18;
+  uint public constant PRICE_CURVE_EXPONENT = 7;
+  uint public constant MAX_PRICE_PERCENTAGE = 1e20;
+
+  /* === MODIFIERS ==== */
+
+  modifier onlyAdvisoryBoard {
+    uint abRole = uint(IMemberRoles.Role.AdvisoryBoard);
+    require(
+      memberRoles().checkRole(msg.sender, abRole),
+      "Cover: Caller is not an advisory board member"
+    );
+    _;
+  }
+
+  constructor(string memory name_, string memory symbol_) ERC721(name_, symbol_) {
+  }
+
+  /* === MUTATIVE FUNCTIONS ==== */
+
+  function buyCover(
+    address owner,
+    uint24 productId,
+    uint8 payoutAsset,
+    uint96 amount,
+    uint32 period,
+    uint maxPrice,
+    StakingPool[] memory stakingPools
+  ) external payable override returns (uint /*coverId*/) {
+    require(initialPrices[productId] != 0, "Cover: product not initialized");
+
+    (uint coverId, uint priceInAsset) = _createCover(owner, productId, payoutAsset, 0, amount, period, stakingPools);
+    require(priceInAsset <= maxPrice, "Cover: Price exceeds maxPrice");
+    retrievePayment(priceInAsset, payoutAsset);
+    return coverId;
+  }
+
+  function createCover(
+    address owner,
+    uint24 productId,
+    uint8 payoutAsset,
+    uint8 deniedClaims,
+    uint96 amount,
+    uint32 period,
+    StakingPool[] calldata stakingPools
+  ) external override returns (uint /*coverId*/) {
+    return 0; // TODO: implement
+  }
+
+  function _createCover(
+    address owner,
+    uint24 productId,
+    uint8 payoutAsset,
+    uint8 deniedClaims,
+    uint96 amount,
+    uint32 period,
+    StakingPool[] memory stakingPools
+  ) internal returns (uint coverId, uint priceInAsset) {
+
+    // convert to NXM amount
+    uint amountToCover = uint(amount) * 1e18 / pool().getTokenPrice(pool().assets(payoutAsset));
+    uint totalPrice = 0;
+
+    for (uint i = 0; i < stakingPools.length; i++) {
+      if (amountToCover == 0) {
+        break;
+      }
+
+      IStakingPool stakingPool = IStakingPool(stakingPools[i].poolAddress);
+      (uint coveredAmount, uint price) = buyCoverFromPool(stakingPool, productId, amountToCover, period);
+      amountToCover -= coveredAmount;
+      totalPrice += price;
+      stakingPoolsForCover[covers.length].push(StakingPool(address(stakingPool), uint96(coveredAmount)));
+    }
+    require(amountToCover == 0, "Not enough available capacity");
+
+    priceInAsset = totalPrice * pool().getTokenPrice(pool().assets(payoutAsset)) / 1e18;
```

Comment:
coverPriceInAsset / coverPriceInCoveredAsset

Reply:
this assumes the decimal count for asset. need to actually know it.